ksp_plugin/vessel_subsets.cpp

```diff
@@ -0,0 +1,71 @@
+#include "ksp_plugin/vessel_subsets.hpp"
+
+#include <list>
+
+#include "ksp_plugin/pile_up.hpp"
+
+namespace principia {
+
+using ksp_plugin::Vessel;
+using ksp_plugin::PileUp;
+
+namespace base {
+
+Subset<Vessel>::Properties::SubsetOfExistingPileUp::SubsetOfExistingPileUp(
+    ContainerIterator<PileUps> pile_up)
+    : pile_up_(pile_up) {
+  missing_ = pile_up_.iterator->vessels().size() - 1;
+}
+
+Subset<Vessel>::Properties::Properties(not_null<ksp_plugin::Vessel*> vessel) {
+  if (vessel->pile_up()) {
+    subset_of_existing_pile_up_.emplace(
+        SubsetOfExistingPileUp(*vessel->pile_up()));
+  }
+  vessels_.emplace_back(vessel);
+}
+
+void Subset<ksp_plugin::Vessel>::Properties::Collect(
+    not_null<PileUps*> pile_ups) {
+  if (!collected_ &&
+      !(subset_of_existing_pile_up_ &&
+        subset_of_existing_pile_up_->missing_ == 0)) {
```

Comment:
This condition is hard to read.  Maybe use variables to give names to some of its parts?